src/lib/plot.typ

```diff
@@ -19,95 +20,71 @@
-    if p == none { return false }
-    let (px, py, ..) = p
-    return (px >= min-x
-        and px <= max-x
-        and py >= min-y
-        and py <= max-y)
-  }
-
-  let lin-interpolated-pt(a, b) = {
-    let x1 = a.at(0)
-    let y1 = a.at(1)
-    let x2 = b.at(0)
-    let y2 = b.at(1)
-
-    /* Special case for vertical lines */
-    if x2 - x1 == 0 {
-      return (x2, calc.min(max-y, calc.max(y2, min-y)))
-    }
-
-    if y2 - y1 == 0 {
-      return (calc.min(max-x, calc.max(x2, min-x)), y2)
-    }
-
-    let m = (y2 - y1) / (x2 - x1)
-    let n = y2 - m * x2
-
-    let x = x2
-    let y = y2
-
-    y = calc.min(max-y, calc.max(y, min-y))
-    x = (y - n) / m
-
-    x = calc.min(max-x, calc.max(x, min-x))
-    y = m * x + n
-
-    return (x, y)
+/// Sample a function with one parameter
+///
+/// - fn (function): Function to sample
+/// - domain (array): X domain tuple (min, max)
+/// - samples (int): Number of samples
+/// - sample-at (array): Sample function at values
+/// -> array: Array of (x, y) points
+#let sample-fn(fn, domain, samples, sample-at: ()) = {
+  assert(samples >= 0,
+    message: "Samples must be positive or zero")
+  assert(samples + sample-at.len() >= 2,
+    message: "You must at least sample 2 values")
+  assert(type(domain) == array and domain.len() == 2,
+    message: "Domain must be a tuple")
+
+  let (lo, hi) = domain
+
+  let y0 = (fn)(lo)
+  let is-vector = type(y0) == array
+  if not is-vector {
+    y0 = ((lo, y0), )
+  } else {
+    y0 = (y0, )
   }
 
-  let paths = ()
+  let pts = sample-at + range(0, samples + 1).map(t => lo + t / samples * (hi - lo))
+  pts = pts.sorted()
 
-  let path = ()
-  let prev-p = none
-  for p in data {
-    if p == none { continue }
-
-    let (px, py, ..) = p
-    if px == none or py == none { continue }
-
-    if in-range(p) {
-      if not in-range(prev-p) and prev-p != none {
-        path.push(lin-interpolated-pt(p, prev-p))
-      }
-
-      path.push(p)
+  return pts.map(x => {
+    if is-vector {
+      (fn)(x)
     } else {
-      if in-range(prev-p) {
-        path.push(lin-interpolated-pt(prev-p, p))
-      } else if prev-p != none {
-        let a = lin-interpolated-pt(p, prev-p)
-        let b = lin-interpolated-pt(prev-p, p)
-        if in-range(a) and in-range(b) {
-          path.push(a)
-          path.push(b)
-        }
-      }
-
-      if path.len() > 0 {
-        paths.push(path)
-        path = ()
-      }
+      (x, (fn)(x))
     }
+  })
+}
 
-    prev-p = p
-  }
-
-  if path.len() > 0 {
-    paths.push(path)
-  }
-  return paths
+/// Sample a function with two parameters
+///
+/// - fn (function): Function to sample
+/// - x-domain (array): X domain tuple (min, max)
+/// - y-domain (array): X domain tuple (min, max)
+/// - x-samples (int): Number of samples
+/// - y-samples (int): Number of samples
+/// - x-sample-at (array): Sample function at x values
+/// - y-sample-at (array): Sample function at y values
+/// -> array: Array of z values
```

Comment:
I know this is an internal function but the parameter documentation barely says anything.

```
Samples the given two parameter function with `x-samples` and `y-samples` values evenly spaced within the range given by `x-domain` and `y-domain` and specific values given by `x-sample-at` and `y-sample-at` and returns each sampled output in an array.

- fn (function): The function to sample. It must have two input parameters of type float, and output a single float/number?.
- x-domain (array): A tuple of numbers that define the domain of the first parameter passed to `fn`. The first item is the minimum, the second item is the maximum. (Is it inclusive (>=) or exclusive (>)?)
- y-domain (array): Same as `x-domain` but for the second parameter passed to `fn`.
- x-samples (int): The number of values to sample the first parameter of `fn` with. The values will be evenly spaced within `x-domain`.
- y-samples (int): Same as `x-samples` but for the second parameter passed to `fn`.
- x-sample-at (array): An array of numbers which should also be used to sample the first parameter of `fn` with. (Should they be between 0 and 1 and then the value is calculated based on the domain? If they are actual values what happens when they are outside of the domain?)
- y-sample-at (array): Same as `x-sample-at` but for the second parameter passed to `fn`.
-> array: The array of outputs of each sample. (how are they orderd?)
```

Reply:
I think both sample fn's should be user facing because you may want to process/analyze sampled data before passing it to the plotting lib, which also accepts the array outputs of said functions as input.